### ISSUE TITLE
Snappy compression performance assessment

### ISSUE THREAD

**Issue by yazzaoui** [autonity/autonity#835]
Snappy compression although designed to be fast and efficient has a significant overhead as shown by the performance profiling report. 
It is unclear if the compression ratio is high enough to be worthy for the kind of data exchange by an autonity node. 

The task would be to assess the bandwidth gains with `snappy.Encode` by monitoring the input and output data size.

https://github.com/autonity/autonity/blob/2f73948bea1c9bb030381fa39127188df5f270c1/p2p/rlpx/rlpx.go#L216-L226

**Comment by vpiyush** [comment]
Snappy Compression results in various lengths of byte arrays.

| Size Before compression   |  Size after Compression  |
|------------------------------------ | ----------------------------------|
| 36 |   38|
122 | 125
197 | 201
1170 | 887
1797 | 1403
2535 | 2006
76807 | 64251

From the above results it’s clear that compression is not effective on smaller buffer sizes, Instead of compressing it adds a few extra bytes. Whereas for the large size buffers (>1k) It does seem to reduce the size **~20%**

We ran a load test on a 40-node setup with 480 TPS.


| |Max CPU | Avg CPU | N/w Ingress(avg) | N/w Egress(avg) |
| ------------- | ------------ | ----------------------- | ---------------------- | ---- |
| With Snappy | 43.87% | 38.35% | 5.38 Mb | 5.65 Mb |
| W/O Snappy | 42.25% | 36.75% | 5.72 Mb | 6.09 Mb |


The same test was then run on an optimized snappy code, There are two major optimizations:
- There is no snappy encoding done on packets smaller than 512 bytes
- The length of the buffer being used for snappy encoding was adjusted before encoding to the actual length needed for encoding.

| | Max CPU | Avg CPU | N/w Ingress(avg) | N/w Egress(avg) |
|-|------------- | ------------ | ----------------------- | ---------------------- | 
| Snappy optimised | 39.25% | 32% | 5.2 Mb | 5.06 Mb |

Here is the CPU profile of all three versions of the code:

Snappy-Enabled
![Snappy-Enabled](https://github.com/autonity/autonity/assets/18163958/6c2f8d31-bf5c-4a7b-b61f-52096955029d)

Snappy-Disabled
![Snappy-Disabled](https://github.com/autonity/autonity/assets/18163958/25229afa-c808-4d8e-a16e-f15f2bcbd3c3)

Snappy-optimized
![Snappy-optimised](https://github.com/autonity/autonity/assets/18163958/71c3ed04-aacc-4b17-8639-d06c776ea519)

**Comment by lorenzo-dev1** [comment]
How do we manage to get less avg cpu usage with snappy optimized wrt without snappy? Is it due to more cpu being used to write a larger amount of data to the network?

**Comment by vpiyush** [comment]
Yes, In the optimized snappy version. If you notice, network bandwidth is ~15% less than the version without snappy. In turn much less data to handle. which would cause low memory allocations too.

**Comment by vpiyush** [comment]
Closing as completed in #843
